### ISSUE TITLE
[PRE REVIEW]: LibSWIFFT - A fast C/C++ Library for the SWIFFT Secure Homomorphic Hash Function

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2962]
**Submitting author:** @yarongvili1 (<a href="http://orcid.org/0000-0002-2221-3598">Yaron Gvili</a>)
**Repository:** <a href="https://github.com/gvilitechltd/LibSWIFFT" target ="_blank">https://github.com/gvilitechltd/LibSWIFFT</a>
**Version:** v1.0.0
**Editor:** @danielskatz
**Reviewers:** @gcdeshpande, @jarrah42
**Managing EiC:** Arfon Smith

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**Author instructions**

Thanks for submitting your paper to JOSS @yarongvili1. **Currently, there isn't an JOSS editor assigned** to your paper.

The author's suggestion for the handling editor is @danielskatz.

@yarongvili1 if you have any suggestions for potential reviewers then please mention them here in this thread (without tagging them with an @). In addition, [this list of people](https://bit.ly/joss-reviewers) have already agreed to review for JOSS and may be suitable for this submission (please start at the bottom of the list).

**Editor instructions**

The JOSS submission bot @whedon is here to help you find and assign reviewers and start the main review. To find out what @whedon can do for you type:

```
@whedon commands
```

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.18 s (169.0 files/s, 17951.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C++                              8            145            281           1558
C/C++ Header                    10             88            310            225
Markdown                         5             85              0            205
C                                4             24            112            144
CMake                            4             20              0             95
-------------------------------------------------------------------------------
SUM:                            31            362            703           2227
-------------------------------------------------------------------------------


Statistical information for the repository '56309c2cc73d3a693460397c' was
gathered on 2021/01/15.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Yaron Gvili                      1          2344              0          100.00

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Yaron Gvili                2344          100.0          0.0               22.35
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2962 with the following error: 

 Can't find any papers to compile :-(

**Comment by arfon** [comment]
@whedon query scope

@yarongvili1 - due to the small size of this code, the editors will now discuss if it meets the [substantial scholarly effort](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort) criterion for review by JOSS. We should get back to you sometime next week.

**Comment by whedon** [comment]
Submission flagged for editorial review.

**Comment by arfon** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02962/joss.02962/10.21105.joss.02962.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.02962/joss.02962/10.21105.joss.02962.pdf) :page_facing_up: :point_left:

**Comment by kthyng** [comment]
@yarongvili1 The editorial board has determined this submission can move forward in the review process. However, it was noted that there seems to be a lack of documentation. We are looking for clear, easy to use docs for a new user to be able to use for installation as well as use of the package. Something like readthedocs is commonly used for this.

**Comment by kthyng** [comment]
@danielskatz you were suggested as a possible editor — would you like to edit this submission?

**Comment by kthyng** [comment]
@whedon invite @danielskatz as editor

**Comment by whedon** [comment]
@danielskatz has been invited to edit this submission.

**Comment by danielskatz** [comment]
@wheden assign me as editor

**Comment by danielskatz** [comment]
@whedon assign me as editor

**Comment by whedon** [comment]
OK, the editor is @danielskatz

**Comment by danielskatz** [comment]
Hi @yarongvili1 - As mentioned by @kthyng, it would be great if you could start improving the documentation before we start the actual review.

**Comment by danielskatz** [comment]
Also, please make suggestions for potential reviewers by mentioning them here in this thread (without tagging them with an @) or by name if they are not on github. [This list of people](https://bit.ly/joss-reviewers) have already agreed to review for JOSS and may be suitable for this submission (please start at the bottom of the list).

**Comment by yarongvili1** [comment]
Hi @kthyng and @danielskatz,

@kthyng, I'd be happy to produce more elaborate documentation for new users. While I'm at it, note that there is existing [quick-start documentation](https://github.com/gvilitechltd/LibSWIFFT/tree/v1.0.0#using-libswifft) as well as [instructions for generating API documentation using doxygen](https://github.com/gvilitechltd/LibSWIFFT/blob/v1.0.0/RELEASE-CHECKLIST.md#running-doxygen).

@danielskatz, a few potential reviewers are: gcdeshpande, olade014, trp07, ceb8.

**Comment by danielskatz** [comment]
Hi @yarongvili1 - sorry for the delay...

Thanks on the documentation - let's see what the reviewers think.

**Comment by danielskatz** [comment]
👋 @gcdeshpande, @trp07, @ceb8, @brnikat, @olade014 - would two of you be willing to review this submission for JOSS?

**Comment by gcdeshpande** [comment]
@danielskatz I can do it.

**Comment by yarongvili1** [comment]
Here is the [issue for the documentation pages](https://github.com/gvilitechltd/LibSWIFFT/issues/1).

**Comment by danielskatz** [comment]
Thanks @gcdeshpande  - I'll assign you, but we won't start the review until we get another reviewer too

**Comment by danielskatz** [comment]
@whedon assign @gcdeshpande as reviewer

**Comment by whedon** [comment]
OK, @gcdeshpande is now a reviewer

**Comment by danielskatz** [comment]
👋 @kevinlewi, @ivmaykov - would either of you be willing to review this submission for JOSS? We carry out our checklist-driven reviews here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html

**Comment by danielskatz** [comment]
Sorry for the delay here...  Let me try again, and also can you potentially suggest any other reviewers?

**Comment by danielskatz** [comment]
👋 @trp07, @ceb8, @Brnikat, @olade014, @kevinlewi, @ivmaykov - would one of you be willing to review this submission for JOSS?  We carry out our checklist-driven reviews here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html

**Comment by danielskatz** [comment]
👋 @holybao, @lukechampine, @s0l0ist - would one of you be willing to review this submission for JOSS? We carry out our checklist-driven reviews here in GitHub issues and follow these guidelines: https://joss.readthedocs.io/en/latest/review_criteria.html

**Comment by gvilitech** [comment]
> Here is the [issue for the documentation pages](https://github.com/gvilitechltd/LibSWIFFT/issues/1).

Resolved. Please see https://libswifft.readthedocs.io

**Comment by yarongvili1** [comment]
> Sorry for the delay here... Let me try again, and also can you potentially suggest any other reviewers?

Not sure. In my suggestion I was aiming for people who say they know cryptography and C/C++.  Maybe security could be an alternative. Perhaps henrykironde, jarrah42, or mahsan321?

**Comment by danielskatz** [comment]
👋 @henrykironde, @jarrah42, @mahsan321 - would one of you be willing to review this submission for JOSS?

**Comment by danielskatz** [comment]
@yarongvili1 and @gcdeshpande - my apologies for the delays in getting this started - I've sent a bunch of emails to potential reviewers, without any positive responses. Given that this is a topic where we don't have a good fit to any of the editors, I decided to take this myself, but this means I also don't have connections to members of the community who might be appropriate reviewers. Do you have any more suggestions for people we could ask?  (even just by name and affiliation?)

**Comment by henrykironde** [comment]
Sorry I did not see this but I am currently reviewing another package. I will not have time to review this one at the same time.

**Comment by danielskatz** [comment]
@henrykironde - do you have any sense when you might be able to review this one? Given the difficulties I'm having in finding reviewers, waiting for you to be available might be a decent option

**Comment by yarongvili1** [comment]
@danielskatz - Unfortunately, I don't have a name off the top of my head that I'm would likely be able to review. I guess let's wait for Henry.

**Comment by jarrah42** [comment]
Sorry for the delay too. This seems to have gotten lost in my GitHub mail. I haven't reviewed before, but I'm willing to take a look.

**Comment by henrykironde** [comment]
I will be free after two weeks. @danielskatz, and @yarongvili1, I have looked at this software and I think the software needs more time before submitting the paper. My recommendation is to withdraw the paper and wait for the software to mature. The concept looks good, but I believe it lacks enough community outreach or utility. I believe it will become a good package. Please take a look at the JOSS criteria. https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort

**Comment by henrykironde** [comment]
One possible alternative to JOSS is to follow [GitHub's guide](https://guides.github.com/activities/citable-code/) on how to create a permanent archive and DOI for your software. This DOI can then be used by others to cite your work

**Comment by Brnikat** [comment]
Sorry not to have responded.

For personal reasons I am not going to be able to take on this task.

Paul

**Comment by danielskatz** [comment]
@henrykironde - when you say "it lacks enough community outreach or utility", can you explain?  The software is a bit small in terms of LOC, but is is sufficient for JOSS to review. Is part of your concern the [substantial scholarly effort](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort) criterion?  A little more on what you think the problem is would be helpful

**Comment by henrykironde** [comment]
@danielskatz, 
1) The number of commits are 9 and the first one was Jan 8 
`As a rule of thumb, JOSS’ minimum allowable contribution should represent not less than three months of work for an individual. Some factors that may be considered by editors and reviewers when judging effort include:`

2) The Software has 19 stars, this is pretty good for the age of the software, however, there is only one fork from the software. Additionally, all issues are from the author, this indicates to me that the software needs more time to gain utility from users

3)  In case this is alongside a contribution describing a science application, details of algorithm development, and/or methods assessment, I expect to see more details on the design of the software, the implementation of the algorithms, creation of tutorials, or any other aspect of the software.
If the software is describing a science application `secure homomorphic hash function`,  I think the author should revise the way they are representing the software. On the other hand, the author mentions the original work, but I do not see any connection between the authors and the works mentioned. If the methods published are not the implementation being submitted to JOSS, I recommend that a team of experts in the field would be needed to review the methods.

> - [The original implementation](https://github.com/micciancio/SWIFFT) written in 2007. It is minimal non-production code. [The AVX2 implementation for K2SN-MSS](https://eprint.iacr.org/2019/442.pdf) is reported to be 25% faster.

**Comment by yarongvili1** [comment]
@henrykironde - Thanks for your quick feedback. Here is some additional information that may be helpful:

1. An important goal of LibSWIFFT (and of many, if not all, implementations of SWIFFT) is high-performance compared to existing implementations and alternative cryptographic hash function. LibSWIFFT is indeed competitive in this respect, as described in the README. The size of the code is not a good indication of the effort needed to attain this goal. Note that while the original implementation is a submission to NIST from 2007 (by other authors), a recent one in K2SN-MSS is from 2019 (also by other authors), and LibSWIFFT is from 2021, so this is still an active topic.
2. Secure hash functions are building blocks in cryptographic applications, rather than complete applications, though still they may be of standalone interest. [Blake3](https://github.com/BLAKE3-team/BLAKE3), which is mentioned in the README, is an example for a well-regarded secure hash function that was implemented in a dedicated GitHub repository.
3. The first commit includes the entire code of v1.0.0 and is the result of several months of work. This work includes finding and fixing a critical bug in the original academic work that relates to the homomorphism property, high-coverage testing including of mathematical properties, adding support for AVX/AVX2/AVX512 instruction sets, and thorough performance optimization, as described in the README file.
4. LibSWIFFT was presented in a lightning talk in Real World Crypto 21' that led to some interest and discussions. It would be good to see LibSWIFFT gain more traction over time, and I believe a JOSS publication could be a part in helping this happen.
5. The full documentation of LibSWIFFT is available at [http://libswifft.readthedocs.io](http://libswifft.readthedocs.io). It includes several guides and a complete API reference.

**Comment by danielskatz** [comment]
Thanks to @henrykironde and @yarongvili1 for this discussion. Overall, the JOSS editors have accepted this for review, so we do want to review it. An outcome of the review could be that the work is not ready to be published, because it doesn't meet the [substantial scholarly effort critereon](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort), but that's not yet clear to me.

Also, while I somewhat agree with some of the points from @henrykironde in terms of best practices, they go beyond JOSS's requirements. There is no JOSS requirement for existing uptake from external users; the main concern is "Reviewers should rely on their expert understanding of their domain to judge whether the software is of broad interest (likely to be cited by other researchers) or more narrowly focused around the needs of an individual researcher or lab."

**Comment by danielskatz** [comment]
@henrykironde - if you are willing to review it against the JOSS criteria in a couple of weeks, that would be great, but if not, I'll continue trying to find another reviewer. Please let me know which you prefer.

**Comment by danielskatz** [comment]
@jarrah42 - sorry, I just saw your message.  Thanks very much!!

I'm going to add you as a reviewer, and start the review.

@henrykironde - if you want to be involved, I can add you as a third reviewer, but if not, that's ok too.

**Comment by danielskatz** [comment]
@whedon add @jarrah42 as reviewer

**Comment by whedon** [comment]
OK, @jarrah42 is now a reviewer

**Comment by danielskatz** [comment]
@whedon start review

**Comment by whedon** [comment]
OK, I've started the review over in https://github.com/openjournals/joss-reviews/issues/3040.

**Comment by henrykironde** [comment]
@danielskatz, yah feel free to add me as a third reviewer

**Comment by danielskatz** [comment]
Thanks @henrykironde - I've now added you over in the review issue: #3040
